### DISCUSSION TITLE
Using custom dimensions from `DynamicQuantities`

### DISCUSSION THREAD

**Discussion by HenryDane** [MilesCranmer/PySR#679] [Q&A] [answered]
Hello,

I'm trying to do equation discovery in a dimensionally consistent way but I'm having a hard time of it. Many of my quantities are either number concentration (#/kg) or mass mixing ratio (kg of A/kg of B). I can not figure out how to express these units using the `PySR` utility. 

I've tried switching to the `SymbolicRegression.jl` library and then implementing my own set of dimensions (following the documentation for `DynamicQuantities`) as show here:

```julia
# ...

# create our dimensions
struct ExtendedDimensions{R} <: AbstractDimensions{R}
    length::R
    mass::R
    time::R
    current::R
    temperature::R
    luminosity::R
    amount::R
    angle::R
    mass_h2o::R
    mass_air::R
    count::R
end

# ... 

const mmr   = Quantity(1.0, ExtendedDimensions(mass_h2o=1, mass_air=-1))
const nconc = Quantity(1.0, ExtendedDimensions(count=1, mass_air=-1))
const dens  = Quantity(1.0, ExtendedDimensions(mass_air=1, length=-3))

X = (; A=data.A * mmr, B=data.B * nconc)
y = data.C * mmr

# create model
model = SRRegressor(
	binary_operators=[*,^,/],
	dimensional_constraint_penalty=10^5,
	nested_constraints=[(^) => [(^) => 0]],
	constraints=[(^) => (-1,1)]
)

# create machine
mach = machine(model, X, y)

# do the fitting
fit!(mach)

r = report(mach)
```

but I get this error when I run the script:

```
ERROR: LoadError: Cannot create a dimensions of `Dimensions{DynamicQuantities.FixedRational{Int32, 25200}}` from `ExtendedDimensions{DynamicQuantities.FixedRational{Int32, 25200}}`. Please write a custom method for construction.
Stacktrace:
  [1] Dimensions{DynamicQuantities.FixedRational{Int32, 25200}}(d::ExtendedDimensions{DynamicQuantities.FixedRational{Int32, 25200}})
    @ DynamicQuantities ~/.julia/packages/DynamicQuantities/5QflN/src/types.jl:118
  [2] convert(::Type{Dimensions{DynamicQuantities.FixedRational{Int32, 25200}}}, d::ExtendedDimensions{DynamicQuantities.FixedRational{Int32, 25200}})
    @ DynamicQuantities ~/.julia/packages/DynamicQuantities/5QflN/src/utils.jl:323
  [3] convert(::Type{Quantity{Float64, Dimensions{DynamicQuantities.FixedRational{Int32, 25200}}}}, q::Quantity{Float64, ExtendedDimensions{DynamicQuantities.FixedRational{Int32, 25200}}})
    @ DynamicQuantities ~/.julia/packages/DynamicQuantities/5QflN/src/utils.jl:314
  [4] get_units(::Type{Float64}, ::Type{Dimensions{DynamicQuantities.FixedRational{Int32, 25200}}}, x::ExtendedDimensions{DynamicQuantities.FixedRational{Int32, 25200}}, ::Function)
    @ SymbolicRegression.InterfaceDynamicQuantitiesModule ~/.julia/packages/SymbolicRegression/RwU5c/src/InterfaceDynamicQuantities.jl:39
  [5] get_si_units(::Type{Float64}, units::ExtendedDimensions{DynamicQuantities.FixedRational{Int32, 25200}})
```

I'm very new to Julia, but it seems to me like the `SRRegressor` needs to "cast" a derived type (`ExtendedDimensions`) to a "base" type (`Dimensions`), and as I understand that, it is not something that can be done. 

Does anyone have any advice? Is it possible to use a custom version of `Dimensions`? 

(I'm asking here since it did not seem reasonable to open an issue on `SymbolicRegression.jl`'s github for this, and the forums link there sent me here).

**Comment by MilesCranmer** [comment, marked as answer]
Hi @HenryDane,

Thanks for the post! Do you want to post this is as an issue on SymbolicRegression.jl? Custom `AbstractDimensions` have actually not yet been set up within `SymbolicRegression` – simply because I didn't know anybody would want to customize this :). But that sounds like a cool use-case, and it should hopefully be fairly easy to add this.

Cheers,
Miles

**Reply by HenryDane** [reply]
I've opened an issue here: https://github.com/MilesCranmer/SymbolicRegression.jl/issues/335

**Comment by HenryDane** [comment]
Sure -- I'll open an issue on the `SymbolicRegression.jl` github.

**Comment by MilesCranmer** [comment]
P.S., as a quick hack, you could treat the other existing units as the units you are interested in. So, e.g., luminosity (cd) could be mass_h20.

**Reply by HenryDane** [reply]
That is a great idea, I'm not sure how I didn't think of that. I'll give it a try!
